mm2src/adex_cli/src/adex_config.rs

```diff
@@ -0,0 +1,178 @@
+
+pub(crate) fn get_config() {
+    let Ok(adex_cfg) = AdexConfigImpl::from_config_path() else { return; };
+    info!("{}", adex_cfg)
+}
+
+pub(crate) fn set_config(set_password: bool, rpc_api_uri: Option<String>) {
+    let mut adex_cfg = AdexConfigImpl::from_config_path().unwrap_or_else(|_| AdexConfigImpl::default());
+    let mut is_changes_happened = false;
+    if set_password {
+        let rpc_password = Password::new("Enter RPC API password:")
+            .prompt()
+            .map(|value| {
+                is_changes_happened = true;
+                value
+            })
+            .map_err(|error| error_anyhow!("Failed to get rpc_api_password: {error}"))
+            .ok();
+        adex_cfg.set_rpc_password(rpc_password);
+    }
+    if rpc_api_uri.is_some() {
+        adex_cfg.set_rpc_uri(rpc_api_uri);
+        is_changes_happened = true;
+    }
+
+    if is_changes_happened && adex_cfg.write_to_config_path().is_ok() {
+        info!("Configuration has been set");
+    } else {
+        warn!("Nothing changed");
+    }
+}
+
+pub(crate) trait AdexConfig {
+    fn rpc_password(&self) -> Result<String>;
+    fn rpc_uri(&self) -> Result<String>;
+    fn orderbook_price_precision(&self) -> &PricePrecision;
+    fn orderbook_volume_precision(&self) -> &VolumePrecision;
+}
+
+#[derive(Deserialize, Serialize, Debug, Default)]
+pub(crate) struct AdexConfigImpl {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    rpc_password: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    rpc_uri: Option<String>,
+}
+
+impl AdexConfig for AdexConfigImpl {
+    fn rpc_password(&self) -> Result<String> {
+        self.rpc_password
+            .as_ref()
+            .map(String::clone)
+            .ok_or_else(|| error_anyhow!("Failed to get rpc_password"))
+    }
+    fn rpc_uri(&self) -> Result<String> {
+        self.rpc_uri
+            .as_ref()
+            .map(String::clone)
+            .ok_or_else(|| error_anyhow!("No rpc_uri in config"))
+    }
+    fn orderbook_price_precision(&self) -> &PricePrecision { &PRICE_PRECISION }
+    fn orderbook_volume_precision(&self) -> &VolumePrecision { &VOLUME_PRECISION }
+}
+
+impl Display for AdexConfigImpl {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        if !self.is_set() {
+            return writeln!(f, "adex configuration is not set");
+        }
+        if let Some(rpc_api_uri) = &self.rpc_uri {
+            writeln!(f, "mm2 RPC URL: {}", rpc_api_uri)?
+        };
+
+        if self.rpc_password.is_some() {
+            writeln!(f, "mm2 RPC password: *************")?
+        }
+        Ok(())
+    }
+}
+
+impl AdexConfigImpl {
+    #[cfg(test)]
+    pub fn new(rpc_password: &str, rpc_uri: &str) -> Self {
+        Self {
+            rpc_password: Some(rpc_password.to_string()),
+            rpc_uri: Some(rpc_uri.to_string()),
+        }
+    }
+
+    #[cfg(not(test))]
+    pub fn read_config() -> Result<AdexConfigImpl> {
```

Comment:
unless you create another `fn read_config` for test environment, you don't need to apply this attribute.

Reply:
Can it be kept as it is?

Reply:
Then you have to do this for 99% of the functions :) It's not going to be included in test environment because you haven't used it anyway

Reply:
Anyway I would have to use allow unused **workaround**. I really don't see the problem in using `#[cfg(test)] ` or `#[cfg(not(test))]`. 

What 99% of the functions do you mean, can you tell me please. Just one example to make me able to understand what do you mean

Reply:
> What 99% of the functions do you mean, can you tell me please. Just one example to make me able to understand what do you mean

Literally almost whole project :D You can take a look in `common` module. Feel free to try `#[cfg(not(test))]` adding this to most of the functions, there will be no difference on the compilation.

> Anyway I would have to use allow unused... workaround.

Hmm, clippy shouldn't care about that. Maybe something is different in there.

Reply:
as an extra fresh example, what I mean is this:

```rs
pub fn hello_there() {}

pub fn hello_there2() {}

fn main() {
    hello_there();
    hello_there2();
    println!("Hello, world!");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_stuff() {
        hello_there2();
    }
}
```

On both environments, there will be no errors. These functions are created in non-test module, so compiler will not expect all of the functions used in the test environment. Imagine if it does, majority of the source code will need to apply this attributes all around.

Reply:
> Hmm, the compiler shouldn't care about that. Maybe something is different in there.

But it cares 

```
warning: associated function `read_config` is never used
   --> src/adex_config.rs:108:12
    |
108 |     pub fn read_config() -> Result<AdexConfigImpl> {
    |            ^^^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted
```

Reply:
> On both environments, there will be no warnings. 

Hm ... there should be warning of coarse